src/httpaf_lwt_client.ml

```diff
@@ -66,34 +68,57 @@ let decode_uri uri =
        (try Ok (host, Some (int_of_string port))
         with Failure _ -> Error (`Msg "couldn't decode port")))
     >>| fun (host, port) ->
-    is_tls, user_pass, host, port, "/" ^ String.concat "/" path
+    is_tls, scheme, user_pass, host, port, "/" ^ String.concat "/" path
   | _ -> Error (`Msg "couldn't decode URI on top")
 
-let prep_headers headers host user_pass blen =
+
+let prep_http_1_1_headers headers host user_pass blen =
+  let headers = Httpaf.Headers.of_list headers in
   let add = Httpaf.Headers.add_unless_exists in
   let headers = add headers "user-agent" ("httpaf-lwt-client/%%VERSION_NUM%%") in
   let headers = add headers "host" host in
   let headers = add headers "connection" "close" in
   let headers = match blen with
     | None -> headers
     | Some x -> add headers "content-length" (string_of_int x)
-  in
+  in 
   match user_pass with
   | None -> headers
   | Some (user, pass) ->
     let data = Base64.encode_string (user ^ ":" ^ pass) in
     let s = "Basic " ^ data in
     add headers "authorization" s
 
-let single_request resolver ?config ?authenticator ~meth ~headers ?body uri =
-  Lwt_result.lift (decode_uri uri) >>= fun (tls, user_pass, host, port, path) ->
-  connect resolver ?port ~tls ?authenticator host >>= fun fd ->
+let prep_h2_headers headers host user_pass blen =
+  let headers = H2.Headers.of_list headers in
+  let add = H2.Headers.add_unless_exists in
+  let headers = add headers ":authority" host in
+  let headers = match blen with
+    | None -> add headers "content-length" "0"
+    | Some x -> add headers "content-length" (string_of_int x)
+  in 
+  match user_pass with
+  | None -> headers
+  | Some (user, pass) ->
+    let data = Base64.encode_string (user ^ ":" ^ pass) in
+    let s = "Basic " ^ data in
+    add headers "authorization" s
```

Comment:
the `user_pass` part is slightly too much code duplication for my taste. also, there are trailing whitespaces and superfluous newlines.